enumerable.rb

```diff
@@ -0,0 +1,169 @@
+  def my_each
+    if block_given?
+      v = self
+      n = v.length
+      n.times do |i|
+        yield(v[i])
+      end
+    else
+      to_enum(:my_each)
+    end
+  end
+
+  def my_each_with_index
+    if block_given?
+      v = self
+      n = v.length
+      n.times do |i|
+        yield(self[i], i)
+      end
+    else
+      to_enum(:my_each_with_index)
+    end
+  end
+
+  def my_select
+    value = self
+    if block_given?
+      temp = []
+      value.my_each do |a|
+        temp.push(a) if yield(a)
+      end
+      temp
+    else
+      to_enum(:my_select)
+    end
+  end
+
+  def my_all?(arg = nil)
+    value = self
+    if block_given?
+      if value.is_a? Hash
+        value.my_each { |a, b| return false unless yield(a, b) }
+      else
+        value.my_each { |a| return false unless yield(a) }
+      end
+    elsif arg.is_a? Regexp
+      value.my_each { |a| return false unless a =~ arg }
+    elsif arg.is_a? Class
+      value.my_each { |a| return false unless a.class == arg }
+    elsif arg.nil?
+      value.my_each { |a| return false unless a }
+    end
+    true
+  end
+
+  def my_any?
+    value = self
+    if block_given?
+      if value.is_a? Hash
+        value.my_each { |a, b| return true if yield(a, b) }
+      else
+        value.my_each { |a| return true if yield(a) }
+      end
+    elsif arg.is_a? Regexp
+      value.my_each { |a| return true if a =~ arg }
+    elsif arg.is_a? Class
+      value.my_each { |a| return true if a.class == arg }
+    elsif arg.nil?
+      value.my_each { |a| return true if a }
+    end
+    false
+  end
+
+  def my_none?
+    value = self
+    if block_given?
+      if value.is_a? Hash
+        value.my_each { |a, b| return false if yield(a, b) }
+      else
+        value.my_each { |a| return false if yield(a) }
+      end
+    elsif arg.is_a? Regexp
+      value.my_each { |a| return false if a =~ arg }
+    elsif arg.is_a? Class
+      value.my_each { |a| return false if a.class == arg }
+    elsif arg.nil?
+      value.my_each { |a| return false if a }
+    end
+    true
+  end
+
+  def my_count(arg = nil)
+    value = self
+    count = 0
+    if block_given?
+      value.my_each do |a|
+        count += 1 if yield(a)
+      end
+    end
+
+    unless arg.nil?
+      value.my_each do |a|
+        count += 1 if arg == a
+      end
+    end
+
+    count = value.length if !block_given? && arg.nil?
+    count
+  end
+
+  def my_map(&block)
+    arr = []
+    value = self
+    if block_given?
+      if value.is_a? Hash
+        value.my_each { |a, b| arr.push(block.call(a, b)) }
+      elsif value.is_a? Array
+        value.my_each { |a| arr.push(block.call(a)) }
+      end
+    else
+      to_enum(:my_map)
+    end
+    arr
+  end
+
+  def my_inject(init = nil, arg = nil, &block)
+    value = self
+    sum
+    if block_given?
+      sum = init.nil? ? value[0] : init
+      value.my_each_with_index { |a, i| sum = block.call(sum, a) unless i.zero? }
```

Comment:
Can we use `value.shift` instead of `value[0]` here and then remove ` unless i.zero? `? :thinking: 

This is because `shift` would have removed the first element when necessary and there won't be a need for excluding the first element in all cases including when it is needed. See example of when the first element is needed:

```ruby
[1,2,3].inject(5){|a,b| a + b}
#=> 11
[1,2,3].my_inject(5){|a,b| a + b}
#=> 10
```

In this example, `init` is not `nil` so we need to include the first element of the array.